test/unit/index/fm_index_iterator_test.cpp

```diff
@@ -0,0 +1,234 @@
+// ============================================================================
+//                 SeqAn - The Library for Sequence Analysis
+// ============================================================================
+//
+// Copyright (c) 2006-2017, Knut Reinert & Freie Universitaet Berlin
+// Copyright (c) 2016-2017, Knut Reinert & MPI Molekulare Genetik
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+#include <type_traits>
+
+#include <gtest/gtest.h>
+
+#include <seqan3/index/fm_index.hpp>
+
+#include <seqan3/alphabet/nucleotide/dna4.hpp>
+
+#include <range/v3/algorithm/equal.hpp>
+
+using namespace seqan3;
+using namespace seqan3::literal;
+
+// NOTE: this test is a typed test for now since we will add bidirectional iterators later
+template <typename T>
+class fm_index_iterator_test : public ::testing::Test
+{};
+
+using fm_index_iterator_types = ::testing::Types<fm_index_iterator<fm_index<std::vector<dna4>>>>;
+
+TYPED_TEST_CASE(fm_index_iterator_test, fm_index_iterator_types);
+
+TYPED_TEST(fm_index_iterator_test, ctr)
+{
+    using text_type = std::vector<dna4>;
+    using index_type = fm_index<std::vector<dna4>>;
+    using iterator_type = TypeParam;
+
+    text_type text {"ACGACG"_dna4};
+    index_type sa{text};
+
+    // custom constructor
```

Comment:
I would split all the tests into parts that test one functionality.
I.e. one test for each constructor, one test for each function.

Reply:
Disagree :smile:

I don't see any need to create a separate test case for each constructor. We could replace testing the default constructors using the metafunctions.